compiler/rustc_parse/src/lib.rs

```diff
@@ -152,14 +152,28 @@ fn try_file_to_source_file(
     path: &Path,
     spanopt: Option<Span>,
 ) -> Result<Lrc<SourceFile>, Diagnostic> {
-    sess.source_map().load_file(path).map_err(|e| {
-        let msg = format!("couldn't read {}: {}", path.display(), e);
+    let mut res = if path.to_string_lossy().ends_with('>') {
+        // The file name can't end with '>' because internally in the compiler
+        // we use the file name "<...>" to represent files that are not actually
+        // real files. To avoid having to split up paths into their components,
```

Comment:
Should we remove that special magic filename `<...>`, and represent fake files without giving them a name?

Reply:
Thank you for reviewing!

Maybe, but what would we use in error messages in that case? Here is an example that emits an error for the file `<anon>`:

```console
$ echo 'fn main_missing() {}' | rustc -
error[E0601]: `main` function not found in crate `rust_out`
 --> <anon>:1:21
  |
1 | fn main_missing() {}
  |                     ^ consider adding a `main` function at the crate level

error: aborting due to previous error

For more information about this error, try `rustc --explain E0601`.
```

What should the error message look like instead do you think?

Even if we want to change it, I don't think we want to change it in this PR? This PR changes an ICE into an error message, which I think is an improvement on its own?

Reply:
The name we use in diagnostics and the name we use in the internal data structure are 2 different things. The issue comes from the fact we hardcode a magic filename in our data structure. We can totally change our data structure, and keep having `<anon>` in the diagnostic.

Reply:
Oh, you mean like that. I agree that would be nice. I'll look into it.

@rustbot author

Reply:
I changed this PR into being about allowing file names to end with '>'. Feel free to take another look.

@rustbot ready